src/tp_engine.hpp

```diff
@@ -157,24 +136,81 @@ class TPEngine {
+  }
+  bool HasLikelihoodEvalEngine() const { return likelihood_engine_ != nullptr; }
+  TPEvaluationEngineViaParsimony &GetParsimonyEvalEngine() {
+    return *parsimony_engine_;
+  }
+  const TPEvaluationEngineViaParsimony &GetParsimonyEvalEngine() const {
+    return *parsimony_engine_;
+  }
+  bool HasParsimonyEvalEngine() const { return parsimony_engine_ != nullptr; }
   EigenConstMatrixXdRef GetLikelihoodMatrix() {
-    return log_likelihoods_.block(0, 0, GetNodeCount(), log_likelihoods_.cols());
+    Assert(HasLikelihoodEvalEngine(), "Must MakeLikelihoodEvalEngine before access.");
+    auto &log_likelihoods =
+        GetLikelihoodEvalEngine().GetDAGBranchHandler().GetBranchLengthData();
+    return log_likelihoods.block(0, 0, GetEdgeCount(), log_likelihoods.cols());
+  }
+  PLVEdgeHandler &GetLikelihoodPVs() {
+    Assert(HasLikelihoodEvalEngine(), "Must MakeLikelihoodEvalEngine before access.");
+    return GetLikelihoodEvalEngine().GetPVs();
+  }
+  PSVEdgeHandler &GetParsimonyPVs() {
+    Assert(HasParsimonyEvalEngine(), "Must MakeParsimonyEvalEngine before access.");
+    return GetParsimonyEvalEngine().GetPVs();
   }
-  PLVEdgeHandler &GetLikelihoodPVs() { return likelihood_pvs_; }
-  PSVEdgeHandler &GetParsimonyPVs() { return parsimony_pvs_; }
-  EigenVectorXd &GetBranchLengths() { return branch_lengths_; }
-  std::vector<size_t> &GetTreeSource() { return tree_source_; }
-  EigenVectorXd &GetTopTreeLikelihoods() { return top_tree_log_likelihoods_per_edge_; }
-  EigenVectorXd &GetTopTreeParsimonies() { return top_tree_parsimony_per_edge_; }
-
-  void SetBranchLengths(EigenVectorXd branch_lengths) {
-    Assert(size_t(branch_lengths.size()) == dag_.EdgeCountWithLeafSubsplits(),
+  EigenVectorXd &GetBranchLengths() {
+    Assert(HasLikelihoodEvalEngine(), "Must MakeLikelihoodEvalEngine before access.");
+    return GetLikelihoodEvalEngine().GetDAGBranchHandler().GetBranchLengthData();
+  }
+  DAGBranchHandler &GetDAGBranchHandler() {
+    Assert(HasLikelihoodEvalEngine(), "Must MakeLikelihoodEvalEngine before access.");
+    return GetLikelihoodEvalEngine().GetDAGBranchHandler();
+  }
+  EigenVectorXd &GetTopTreeLikelihoods() {
+    Assert(HasLikelihoodEvalEngine(), "Must MakeLikelihoodEvalEngine before access.");
+    return GetLikelihoodEvalEngine().GetTopTreeScores();
+  }
+  EigenVectorXd &GetTopTreeParsimonies() {
+    Assert(HasParsimonyEvalEngine(), "Must MakeParsimonyEvalEngine before access.");
+    return GetParsimonyEvalEngine().GetTopTreeScores();
+  }
+  void SetBranchLengths(EigenVectorXd new_branch_lengths) {
+    Assert(HasLikelihoodEvalEngine(), "Must MakeLikelihoodEvalEngine before access.");
+    auto &branch_lengths = GetBranchLengths();
+    Assert(size_t(new_branch_lengths.size()) == GetEdgeCount(),
            "Size mismatch in GPEngine::SetBranchLengths.");
-    branch_lengths_.segment(0, dag_.EdgeCountWithLeafSubsplits()) = branch_lengths;
+    for (size_t i = 0; i < GetEdgeCount(); i++) {
```

Comment:
Seems like this could be done via a BranchHandler setter, right?

Reply:
Yes, changed.